core/src/entry_verification_service/worker.rs

```diff
@@ -0,0 +1,524 @@
+        /// next, when updating [`SlotStatus::verified_up_to`].
+        next_data_set: DataSetFirstShred,
+        /// Hash of the very last entry.  Necessary for the PoH verification of the first entry in
+        /// the next data set.
+        last_entry_hash: Hash,
+    },
+}
+
+impl DataSetOptimisticStatus {
+    fn last_entry_hash(&self) -> Option<Hash> {
+        match self {
+            Self::PohStartHash(_) => None,
+            Self::AllButFirst {
+                last_entry_hash, ..
+            } => Some(*last_entry_hash),
+            Self::Verified {
+                last_entry_hash, ..
+            } => Some(*last_entry_hash),
+        }
+    }
+}
+
+struct State {
+    commands: crossbeam_channel::Receiver<Command>,
+
+    blockstore: Arc<Blockstore>,
+
+    processing: HashMap<Slot, SlotStatus>,
+
+    /// Slots that are waiting for the parent slot last entry hash.
+    waiting_for_parent_hash: HashMap<Slot, Vec<Slot>>,
+
+    /// When a client is actively waiting for a certain data set that we did not process yet, we are
+    /// going to skip accumulation in order to provide better latency.
+    ///
+    /// We skip accumulation if we see any work for a data set in this slot, for any of the data
+    /// sets up to and including the specified one.  We also skip accumulation, if we see any work
+    /// for a parent slot.
+    ///
+    /// As slots are relatively big compared to datasets, it is unlikely that the validation service
+    /// will be waiting accumulating data for a grand-parent of the slot the clients are waiting
+    /// for.
+    client_waiting_for: Option<(Slot, DataSetFirstShred)>,
+
+    results: Arc<SharedResults>,
+}
+
+/// Indicates which validation are we planning to do for a data set.
+///
+/// As we are going to run verification for PoH and for signature verification in different threads,
+/// and, potentially, in parallel, holding `entries` as an `Arc<Vec<Entry>>` saves an extra copy.
+#[derive(PartialEq, Eq, Debug, Clone)]
+enum TodoOp {
+    /// Verify PoH for entries, and signature verification for all transactions.
+    All {
+        poh_start_hash: Hash,
+        entries: Arc<Vec<Entry>>,
+    },
+    /// PoH for all but the first entry, and signature verification for all transactions.
+    AllButPohOnFirstEntry { entries: Arc<Vec<Entry>> },
+    /// PoH for the first entry, only.
+    OnlyFirstEntryPoh { poh_start_hash: Hash, entry: Entry },
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+struct TodoItem {
+    /// Slot to mark as failed if the verification specified in `op` fails.
+    slot: Slot,
+    /// Pending verification operation.
+    op: TodoOp,
+}
+
+type Todo = Vec<TodoItem>;
+
+impl State {
+    fn run(&mut self) {
+        loop {
+            let todo = match self.accumulate() {
+                Some(todo) => todo,
+                None => return,
+            };
+
+            self.verify(todo);
+        }
+    }
+
+    fn accumulate(&mut self) -> Option<Todo> {
```

Comment:
maybe `accumulate_verification_requests` would be more explicit

Reply:
This is a local function that is only called by `run()`.
For such a small scope, it seems nice to use short names.
`run()` is just
```rust
        loop {
            let todo = match self.accumulate() {
                Some(todo) => todo,
                None => return,
            };

            self.verify(todo);
        }
```

Everything is about verification here, as this is a verification service :)
Maybe `accumulate_work()` could be a clarification?
You are right that `accumulate()` by itself could refer to many different things %)
What do you think?

Reply:
`accumulate_work()` works too. We use accumulate a lot of places for metrics, so I'll be happy with any postfix to save my future self :)

Reply:
Done.